R/cube-result.R

```diff
@@ -13,9 +13,249 @@ setMethod("initialize", "CrunchCube", function (.Object, ...) {
     return(.Object)
 })
 
+# ---- Cube Subset ----
+
+#' @rdname cube-methods
+#' @export
+setMethod("[", "CrunchCube", function (x, i, j, ..., drop = TRUE) {
+    subset <- eval(substitute(alist(i, j, ...)))
+    subset <- replaceMissingWithTRUE(subset)
+    dims <- dim(x)
+    if (length(subset) != length(dims)) {
+        halt("You must supply ",
+            length(subset),
+            " dimensions to subset a ",
+            length(dims),
+            " dimensional cube.")
+    }
+
+    err_subsets <- lapply(subset, function(idx) {
+        if (is.numeric(idx)) {
+            return(max(idx))
+        } else if (is.logical(idx) && !isTRUE(idx)) {
+            return(length(idx))
+        }
+        # Assume other subsets are valid and trust that the array subsetting
+        # method will fail if they are invalid.
+        return (TRUE)
+    })
+
+    isInValid <- function(idx, dim) {
+        if (isTRUE(idx)) {
+            return(FALSE)
+        }
+        return(idx > dim)
+    }
+
+    invalid_subsets <- unlist(mapply(isInValid, idx = err_subsets, dim = dims, SIMPLIFY = logical(1)))
+    if (any(invalid_subsets)) {
+        errs <- paste("- At position", which(invalid_subsets),
+            "you tried to select element",
+            err_subsets[invalid_subsets],
+            "when the dimension has",
+            dims[invalid_subsets],
+            "elements."
+        )
+        halt("Invalid subset:\n",
+            paste0(errs[1:min(sum(invalid_subsets), 5)], collapse = "\n"))
+    }
+
+    translated_subset <- translateCubeIndex(x, subset, drop)
+    if (x@useNA == "no") {
+        translated_subset <- skipMissingCategories(x, translated_subset)
+    }
+
+    out <- x
+    out@arrays[] <- lapply(out@arrays, function(arr){
+        subsetByList(arr, translated_subset, drop)
+    })
+
+    out@dims[] <- mapply(subsetArrayDimension,
+        dim = x@dims,
+        idx = translated_subset,
+        SIMPLIFY = FALSE)
+
+    if (drop) {
+        keep_args <- vapply(translated_subset, function(a) {
+            if (is.logical(a)) {
+                return(sum(a) != 1 || isTRUE(a))
+            } else {
+                return(length(a) != 1)
+            }
+        }, FUN.VALUE = logical(1))
+        out@dims <- out@dims[keep_args]
+    }
+    return(out)
+})
+
+#' Replace missing elements with TRUE
+#'
+#' When subsetting missingness stands in for selecting all of the elements of that
+#' dimension. This gets tricky when you are selecting n dimensions and so need to
+#' capture the dimensions with `...`.  This function checks if an element of a list
+#' is missing and replaces that element with `TRUE` in order to handle this case.
+#'
+#' @param l a list
+#' @return a list
+#' @keywords internal
+replaceMissingWithTRUE <- function(l){
+    out <- lapply(l, function(x){
+        if (is.symbol(x)) {
```

Comment:
This works, and I think is safe, but I'd still prefer something cleaner for this test. There is a great function in `rlang` which does just the trick, if we want to include that dependency. https://www.rdocumentation.org/packages/rlang/versions/0.1.6/topics/missing_arg